################################################################################
##                                                                            ##
## Plot scores from one method against those of all others                    ##
##                                                                            ##
## Inputs:                                                                    ##
## * scorerds: list containing abundance estimates and characteristics for    ##
##             junctions, transcripts and genes, as well as gene scores.      ##
##             Generated by calculate_gene_scores.R                           ##
## * targetmethod: method to which all other methods will be compared         ##
## * uniqjuncreadsthreshold: the total number of uniquely mapping junction    ##
##                           reads (in a gene), only genes with more than     ##
##                           this number will be used for the comparison      ##
## * outrds: output rds file. The name will be used to determine the name of  ##
##           the output figures.                                              ##
##                                                                            ##
## Outputs:                                                                   ##
## * A png figure comparing the scores of all methods to those of the target  ##
##   method                                                                   ##
## * A png figure summarizing the number of genes for which the target method ##
##   gives higher/lower/equal score compared to each other method             ##
## * For each method, a text file with the genes for which the target method  ##
##   gave a lower score                                                       ##
##                                                                            ##
################################################################################

args <- (commandArgs(trailingOnly = TRUE))
for (i in 1:length(args)) {
  eval(parse(text = args[[i]]))
}

print(scorerds)
print(targetmethod)
print(uniqjuncreadsthreshold)
print(outrds)

suppressPackageStartupMessages({
  library(dplyr)
  library(ggplot2)
  library(cowplot)
})

scores <- readRDS(scorerds)
scores <- scores$genes

## Consider only genes with uniqjuncreads > uniqjuncreadsthreshold
tmpscore <- scores %>% dplyr::filter(uniqjuncreads > uniqjuncreadsthreshold) %>% 
  dplyr::select(gene, method, score)
scoresub <- tmpscore %>%
  tidyr::spread(method, score) %>% as.data.frame() %>% 
  tibble::column_to_rownames("gene")

png(gsub("rds$", "png", outrds), width = 8, height = 8, unit = "in", res = 300)
plots <- lapply(setdiff(colnames(scoresub), targetmethod), function(cn) {
  ggplot(scoresub, aes_string(x = cn, y = targetmethod)) + 
    geom_abline(intercept = 0, slope = 1) + 
    geom_point(alpha = 0.3, size = 0.3) + theme_bw()
})
print(cowplot::plot_grid(plotlist = plots))
dev.off()

scorecomp <- tmpscore %>%
  dplyr::left_join(tmpscore %>% dplyr::filter(method == targetmethod) %>%
                     dplyr::select(gene, score) %>%
                     dplyr::rename(targetscore = score)) %>%
  dplyr::filter(method != targetmethod) %>%
  dplyr::group_by(method) %>%
  dplyr::summarize(ImNmorethanp5 = sum(score - targetscore >= 0.5, na.rm = TRUE),
                   ImNbtw0andp5 = sum(score - targetscore > 0 & score - targetscore < 0.5, 
                                     na.rm = TRUE),
                   ImN0 = sum(score == targetscore, na.rm = TRUE),
                   NmIbtw0andp5 = sum(score - targetscore < 0 & score - targetscore > (-0.5), 
                                     na.rm = TRUE),
                   NmImorethanp5 = sum(score - targetscore <= (-0.5), na.rm = TRUE)) %>%
  tidyr::gather(category, nbr_genes, -method) %>%
  dplyr::mutate(category = replace(category, category == "ImN0", sprintf("score = %s score", targetmethod)),
                category = replace(category, category == "ImNbtw0andp5", sprintf("0 < score - %s score < 0.5", 
                                                                                 targetmethod)),
                category = replace(category, category == "ImNmorethanp5", sprintf("score - %s score >= 0.5",
                                                                                  targetmethod)),
                category = replace(category, category == "NmIbtw0andp5", sprintf("0 < %s score - score < 0.5",
                                                                                 targetmethod)),
                category = replace(category, category == "NmImorethanp5", sprintf("%s score - score >= 0.5",
                                                                                  targetmethod))) %>%
  dplyr::mutate(category = factor(category, levels = c(sprintf("%s score - score >= 0.5", targetmethod), 
                                                       sprintf("0 < %s score - score < 0.5", targetmethod),
                                                       sprintf("score = %s score", targetmethod),
                                                       sprintf("0 < score - %s score < 0.5", targetmethod),
                                                       sprintf("score - %s score >= 0.5", targetmethod))))
png(gsub("\\.rds$", "_summary.png", outrds), width = 8, height = 8,
    unit = "in", res = 300)
ggplot(scorecomp, aes(x = method, y = nbr_genes, fill = category)) + 
  geom_bar(stat = "identity") + 
  scale_fill_manual(values = structure(c("#000099", "#8080ff", "#ffc6b3",
                                         "#ff6633", "#992600"), 
                                       names = c(sprintf("%s score - score >= 0.5", targetmethod),
                                                 sprintf("0 < %s score - score < 0.5", targetmethod),
                                                 sprintf("score = %s score", targetmethod),
                                                 sprintf("0 < score - %s score < 0.5", targetmethod),
                                                 sprintf("score - %s score >= 0.5", targetmethod))), 
                    name = "") + 
  theme_bw() + xlab("") + ylab("Number or genes") + 
  guides(fill = guide_legend(nrow = 3)) + theme(legend.position = "bottom")
dev.off()

## Write genes with lower score with the target method to files
for (cn in setdiff(colnames(scoresub), targetmethod)) {
  tmp <- scoresub[order(scoresub[, targetmethod] - scoresub[, cn]), ]
  genes_to_write <- rownames(tmp)[which(tmp[, targetmethod] < tmp[, cn])]
  write.table(scores %>% dplyr::filter(gene %in% genes_to_write & 
                                         method %in% c(cn, targetmethod)) %>%
                dplyr::select(gene, method, count, length_diff_3putrs_samestart,
                              exoncount, introncount, intron_exon_ratio,
                              uniqjuncreads, mmjuncreads, uniqjuncfraction,
                              nbr_junctions_in_gene, score) %>%
                tidyr::gather(variable, value, count, score) %>%
                tidyr::unite(temp, method, variable) %>%
                tidyr::spread(temp, value) %>%
                dplyr::mutate(gene = factor(gene, levels = genes_to_write)) %>%
                dplyr::arrange(gene),
              file = gsub("\\.rds$", paste0("_", targetmethod, "_lowerthan_", cn, ".txt"), outrds),
              row.names = FALSE, col.names = TRUE, quote = FALSE, sep = "\t")
}

saveRDS(NULL, file = outrds)
date()
sessionInfo()
