################################################################################
##                                                                            ##
## Plot performance for simulated data                                        ##
##                                                                            ##
## Inputs:                                                                    ##
## * scorerds: list containing abundance estimates and characteristics for    ##
##             junctions, transcripts and genes, as well as gene scores.      ##
##             Generated by calculate_gene_scores.R                           ##
## * truthrda: the file generated by polyester, containing true transcript    ##
##             abundances (or '' if not a simulated data set)                 ##
## * truthmodgenesrds: vector with the names of the genes whose transcripts   ##
##                     have been altered (or '' if not a simulated data set)  ##
## * gtf: gtf file, to get 3'UTR lengths                                      ##
## * uniqjuncreadsthreshold: the total number of uniquely mapping junction    ##
##                           reads (in a gene), only genes with more than     ##
##                           this number will be used for the comparison      ##
## * outrds: output file                                                      ##
##                                                                            ##
## Outputs:                                                                   ##
## * png figures illustrating the relative contribution of each type of       ##
##   transcript (contributing UTR, contributing internal structure, other)    ##
##   to the count and TPM of a gene                                           ##
##                                                                            ##
################################################################################

args <- (commandArgs(trailingOnly = TRUE))
for (i in 1:length(args)) {
  eval(parse(text = args[[i]]))
}

print(scorerds)
print(truthrda)
print(truthmodgenesrds)
print(gtf)
print(uniqjuncreadsthreshold)
print(outrds)

suppressPackageStartupMessages({
  library(rtracklayer)
  library(dplyr)
  library(tidyr)
  library(ggplot2)
  library(GenomicRanges)
  library(GenomicFeatures)
  library(cowplot)
})

scores <- readRDS(scorerds)
load(truthrda)
modgenes <- readRDS(truthmodgenesrds)
gtf <- rtracklayer::import(gtf, format = "gtf")

transcripts <- scores$transcripts

## Get the coordinates and lengths of the 3'UTR for each transcript
utrs <- as.data.frame(gtf) %>% dplyr::filter(type == "three_prime_utr") %>%
  dplyr::group_by(transcript_id) %>%
  dplyr::summarize(start = min(start), end = max(end), width = sum(width), 
                   seqnames = seqnames[1], gene_id = gene_id[1], strand = strand[1])

## Get the total length of each transcript
exonunions <- as.data.frame(gtf) %>% dplyr::filter(type == "exon") %>%
  dplyr::group_by(transcript_id) %>%
  dplyr::summarize(start = min(start), end = max(end), width = sum(width),
                   seqnames = seqnames[1], gene_id = gene_id[1], strand = strand[1])

## Set up data frame with information about modified transcripts.
## utr_fraction_of_length = fraction of artificial transcript that is made up of
## the 3'UTR 
## utr_selected = "short" or "long", depending on which of the two
## 3'UTRs that was used for the artificial transcript
modtrans <- data.frame(
  mod_transcript = grep("utrfrom", rownames(counts_matrix), value = TRUE), 
  stringsAsFactors = FALSE) %>%
  tidyr::separate(mod_transcript, into = c("internal_tx", "utr_tx"), 
                  sep = "_utrfrom_", remove = FALSE) %>%
  dplyr::mutate(internal_tx_utrlength = utrs$width[match(internal_tx, utrs$transcript_id)],
                utr_tx_utrlength = utrs$width[match(utr_tx, utrs$transcript_id)]) %>%
  dplyr::mutate(gene = transcripts$gene[match(internal_tx, transcripts$transcript)]) %>%
  dplyr::mutate(utr_selected = ifelse(internal_tx_utrlength > utr_tx_utrlength, "short", "long")) %>%
  dplyr::mutate(utr_fraction_of_length = utr_tx_utrlength/(exonunions$width[match(internal_tx, exonunions$transcript_id)] - internal_tx_utrlength + utr_tx_utrlength))

png(gsub("\\.rds", "_utr_fraction_of_length.png", outrds), width = 6, height = 6,
    unit = "in", res = 300)
print(ggplot(modtrans, aes(x = utr_selected, y = utr_fraction_of_length)) + 
        geom_boxplot(outlier.size = 0.75) + theme_bw() + 
        xlab("Selected 3'UTR for artificial transcript") + 
        ylab("Fraction of artificial transcript consisting of 3'UTR") + 
        theme(axis.text = element_text(size = 13),
              axis.title = element_text(size = 15)))
dev.off()
      
## Generate GRangesList with one entry per artificial transcript. This will be
## used to find the most similar annotated transcript to each artificial
## transcript.
txdb <- makeTxDbFromGRanges(gtf)
ebt <- exonsBy(txdb, "tx", use.names = TRUE)
utrs0 <- subset(gtf, type == "three_prime_utr")
L <- as(lapply(seq_len(nrow(modtrans)), function(i) {
  ## Full UTR-contributing transcript
  utrtx <- ebt[[modtrans$utr_tx[i]]]
  ## Full internal structure-contributing transcript
  inttx <- ebt[[modtrans$internal_tx[i]]]
  ## 3'UTR of UTR-contributing transcript
  utrutrtx <- subset(utrs0, transcript_id == modtrans$utr_tx[i])
  ## 3'UTR of internal structure-contributing transcript
  utrinttx <- subset(utrs0, transcript_id == modtrans$internal_tx[i])
  GenomicRanges::reduce(GenomicRanges::union(GenomicRanges::setdiff(inttx, utrinttx), utrutrtx))
}), "GRangesList")
names(L) <- modtrans$mod_transcript

## Find overlaps between modified transcripts and annotated ones
ol <- findOverlaps(L, ebt)

## Find most similar transcript for each artificial transcript
jaccards <- do.call(rbind, lapply(seq_len(length(ol)), function(i) {
  pin <- GenomicRanges::intersect(L[[queryHits(ol)[i]]], ebt[[subjectHits(ol)[i]]])
  pun <- GenomicRanges::union(L[[queryHits(ol)[i]]], ebt[[subjectHits(ol)[i]]])
  overlap <- sum(width(pin))/sum(width(pun))
  data.frame(artificial_tx = queryHits(ol)[i], reference_tx = subjectHits(ol)[i], jaccard = overlap)
}))
jaccards <- jaccards %>% 
  dplyr::mutate(artificial_tx = names(L)[artificial_tx],
                reference_tx = names(ebt)[reference_tx]) %>%
  dplyr::mutate(modified_gene = transcripts$gene[match(sapply(strsplit(artificial_tx, "_utrfrom"), 
                                                              .subset, 1), transcripts$transcript)],
                reference_gene = transcripts$gene[match(reference_tx, transcripts$transcript)]) %>%
  dplyr::filter(modified_gene == reference_gene) %>% 
  dplyr::group_by(artificial_tx) %>% 
  dplyr::summarize(reference_tx = reference_tx[which.max(jaccard)], 
                   jaccard = jaccard[which.max(jaccard)])

summary(jaccards$jaccard)

## Remove transcripts that are identical to an annotated transcript from the
## list of modified transcripts
modtrans <- subset(modtrans, mod_transcript %in% jaccards$artificial_tx[jaccards$jaccard < 1])

## Subset transcript abundance table to only the transcripts in the modified genes
gene_summary <- transcripts %>% 
  dplyr::filter(gene %in% modtrans$gene) %>%
  dplyr::mutate(tr_type = ifelse(transcript %in% modtrans$internal_tx, 
                                 "Contributing internal structure", 
                                 ifelse(transcript %in% modtrans$utr_tx, 
                                        "Contributing 3'UTR", "Other"))) %>%
  dplyr::mutate(tr_type2 = ifelse(transcript %in% jaccards$reference_tx, 
                                  "Most similar reference transcript", "Other")) %>% 
  dplyr::left_join(modtrans %>% dplyr::select(gene, utr_selected, utr_fraction_of_length)) %>%
  dplyr::select(transcript, gene, count, TPM, tr_type, utr_selected, method, 
                tr_type2, utr_fraction_of_length)

sink(file = gsub("\\.rds$", "_transcript_types.txt", outrds))
table(gene_summary$tr_type, gene_summary$tr_type2, gene_summary$method)
sink()

## Plot relative contributions to gene count/TPM from the different transcript
## categories
png(gsub("\\.rds$", "_count.png", outrds), width = 10, height = 7, unit = "in", res = 300)
print(ggplot(gene_summary %>%
               dplyr::group_by(gene, utr_selected, tr_type, method, utr_fraction_of_length) %>% 
               dplyr::summarize(count = sum(count), TPM = sum(TPM)) %>% dplyr::ungroup() %>%
               dplyr::group_by(gene, method) %>% dplyr::mutate(count = count/sum(count),
                                                               TPM = TPM/sum(TPM)) %>%
               dplyr::ungroup(), 
             aes(x = utr_selected, y = count, color = tr_type)) + 
        geom_boxplot(outlier.size = 0.3) + facet_wrap(~ method, nrow = 2) + 
        theme_bw() + scale_color_manual(values = c("#9900cc", "#009933", "#0099cc"),
                                        name = "Transcript class") + 
        xlab("Selected 3'UTR") + ylab("Relative contribution to gene count") + 
        theme(legend.position = "bottom"))
dev.off()

png(gsub("\\.rds$", "_count2.png", outrds), width = 10, height = 7, unit = "in", res = 300)
print(ggplot(gene_summary %>%
               dplyr::group_by(gene, utr_selected, tr_type2, method, utr_fraction_of_length) %>% 
               dplyr::summarize(count = sum(count), TPM = sum(TPM)) %>% dplyr::ungroup() %>%
               dplyr::group_by(gene, method) %>% dplyr::mutate(count = count/sum(count),
                                                               TPM = TPM/sum(TPM)) %>%
               dplyr::ungroup(), 
             aes(x = tr_type2, y = count, color = tr_type2)) + 
        geom_boxplot(outlier.size = 0.3) + facet_wrap(~ method, nrow = 2) + 
        theme_bw() + scale_color_manual(values = c("#9900cc", "#009933"),
                                        name = "") + 
        xlab("") + ylab("Relative contribution to gene count") + 
        theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
              legend.position = "none"))
dev.off()

g1 <- gene_summary %>% dplyr::filter(tr_type2 == "Most similar reference transcript") %>%
  dplyr::mutate(most_similar = tr_type)
g0 <- gene_summary %>% dplyr::left_join(g1 %>% dplyr::select(gene, method, most_similar)) %>%
  dplyr::group_by(gene, utr_selected, tr_type, method, utr_fraction_of_length, most_similar) %>% 
  dplyr::summarize(count = sum(count), TPM = sum(TPM)) %>% dplyr::ungroup() %>%
  dplyr::group_by(gene, method) %>% dplyr::mutate(count = count/sum(count),
                                                  TPM = TPM/sum(TPM)) %>%
  dplyr::ungroup() %>% dplyr::filter(!is.na(most_similar)) %>%
  dplyr::mutate(most_similar = replace(most_similar, most_similar == "Contributing 3'UTR", "3'UTR most sim"),
                most_similar = replace(most_similar, most_similar == "Contributing internal structure", "Internal most sim"),
                most_similar = replace(most_similar, most_similar == "Other", "Other most sim"))
png(gsub("\\.rds$", "_count3.png", outrds), width = 12, height = 8, unit = "in", res = 300)
plots <- lapply(split(g0, g0$method), function(g) {
  ggplot(g, 
         aes(x = utr_selected, y = count, color = tr_type)) + 
    geom_boxplot(outlier.size = 0.3) + facet_grid(most_similar ~ method) + 
    theme_bw() + scale_color_manual(values = c("#9900cc", "#009933", "#0099cc"),
                                    name = "Transcript class") + 
    xlab("Selected 3'UTR") + ylab("Relative contribution to gene count") + 
    theme(legend.position = "bottom")
})
print(cowplot::plot_grid(
  cowplot::plot_grid(plotlist = lapply(plots, function(p) p + theme(legend.position = "none")), nrow = 2),
  cowplot::get_legend(plots[[1]]), ncol = 1, rel_heights = c(1, 0.05))
)
dev.off()

fun_length <- function(x){
  return(data.frame(y = 1.1, label = paste0("n=", length(x))))
}

png(gsub("\\.rds$", "_count4.png", outrds), width = 12, height = 8, unit = "in", res = 300)
plots <- lapply(split(g0, g0$method), function(g) {
  ggplot(g, 
         aes(x = utr_selected, y = count, color = tr_type)) + 
    geom_boxplot(outlier.size = 0.3) + facet_grid(most_similar ~ method) + 
    theme_bw() + scale_color_manual(values = c("#9900cc", "#009933", "#0099cc"),
                                    name = "Transcript class") + 
    xlab("Selected 3'UTR") + ylab("Relative contribution to gene count") + 
    theme(legend.position = "bottom") + 
    stat_summary(position = position_dodge(0.75),
                 fun.data = fun_length, geom = "text",
                 vjust = 0, size = 3) + 
    expand_limits(y = 1.25)
})
print(cowplot::plot_grid(
  cowplot::plot_grid(plotlist = lapply(plots, function(p) p + theme(legend.position = "none")), nrow = 2),
  cowplot::get_legend(plots[[1]]), ncol = 1, rel_heights = c(1, 0.05))
)
dev.off()


## TPMs
png(gsub("\\.rds$", "_tpm.png", outrds), width = 10, height = 7, unit = "in", res = 300)
print(ggplot(gene_summary %>%
               dplyr::group_by(gene, utr_selected, tr_type, method, utr_fraction_of_length) %>% 
               dplyr::summarize(count = sum(count), TPM = sum(TPM)) %>% dplyr::ungroup() %>%
               dplyr::group_by(gene, method) %>% dplyr::mutate(count = count/sum(count),
                                                               TPM = TPM/sum(TPM)) %>%
               dplyr::ungroup(), 
             aes(x = utr_selected, y = TPM, color = tr_type)) + 
        geom_boxplot(outlier.size = 0.3) + facet_wrap(~ method, nrow = 2) + 
        theme_bw() + scale_color_manual(values = c("#9900cc", "#009933", "#0099cc"),
                                        name = "Transcript class") + 
        xlab("Selected 3'UTR") + ylab("Relative contribution to gene TPM") + 
        theme(legend.position = "bottom"))
dev.off()

png(gsub("\\.rds$", "_tpm2.png", outrds), width = 10, height = 7, unit = "in", res = 300)
print(ggplot(gene_summary %>%
               dplyr::group_by(gene, utr_selected, tr_type2, method, utr_fraction_of_length) %>% 
               dplyr::summarize(count = sum(count), TPM = sum(TPM)) %>% dplyr::ungroup() %>%
               dplyr::group_by(gene, method) %>% dplyr::mutate(count = count/sum(count),
                                                               TPM = TPM/sum(TPM)) %>%
               dplyr::ungroup(), 
             aes(x = tr_type2, y = TPM, color = tr_type2)) + 
        geom_boxplot(outlier.size = 0.3) + facet_wrap(~ method, nrow = 2) + 
        theme_bw() + scale_color_manual(values = c("#9900cc", "#009933"),
                                        name = "") + 
        xlab("") + ylab("Relative contribution to gene TPM") + 
        theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
              legend.position = "none"))
dev.off()

g1 <- gene_summary %>% dplyr::filter(tr_type2 == "Most similar reference transcript") %>%
  dplyr::mutate(most_similar = tr_type)
g0 <- gene_summary %>% dplyr::left_join(g1 %>% dplyr::select(gene, method, most_similar)) %>%
  dplyr::group_by(gene, utr_selected, tr_type, method, utr_fraction_of_length, most_similar) %>% 
  dplyr::summarize(count = sum(count), TPM = sum(TPM)) %>% dplyr::ungroup() %>%
  dplyr::group_by(gene, method) %>% dplyr::mutate(count = count/sum(count),
                                                  TPM = TPM/sum(TPM)) %>%
  dplyr::ungroup() %>% dplyr::filter(!is.na(most_similar)) %>%
  dplyr::mutate(most_similar = replace(most_similar, most_similar == "Contributing 3'UTR", "3'UTR most sim"),
                most_similar = replace(most_similar, most_similar == "Contributing internal structure", "Internal most sim"),
                most_similar = replace(most_similar, most_similar == "Other", "Other most sim"))
png(gsub("\\.rds$", "_tpm3.png", outrds), width = 12, height = 8, unit = "in", res = 300)
plots <- lapply(split(g0, g0$method), function(g) {
  ggplot(g, 
         aes(x = utr_selected, y = TPM, color = tr_type)) + 
    geom_boxplot(outlier.size = 0.3) + facet_grid(most_similar ~ method) + 
    theme_bw() + scale_color_manual(values = c("#9900cc", "#009933", "#0099cc"),
                                    name = "Transcript class") + 
    xlab("Selected 3'UTR") + ylab("Relative contribution to gene TPM") + 
    theme(legend.position = "bottom")
})
print(cowplot::plot_grid(
  cowplot::plot_grid(plotlist = lapply(plots, function(p) p + theme(legend.position = "none")), nrow = 2),
  cowplot::get_legend(plots[[1]]), ncol = 1, rel_heights = c(1, 0.05))
)
dev.off()

fun_length <- function(x){
  return(data.frame(y = 1.1, label = paste0("n=", length(x))))
}

png(gsub("\\.rds$", "_tpm4.png", outrds), width = 12, height = 8, unit = "in", res = 300)
plots <- lapply(split(g0, g0$method), function(g) {
  ggplot(g, 
         aes(x = utr_selected, y = TPM, color = tr_type)) + 
    geom_boxplot(outlier.size = 0.3) + facet_grid(most_similar ~ method) + 
    theme_bw() + scale_color_manual(values = c("#9900cc", "#009933", "#0099cc"),
                                    name = "Transcript class") + 
    xlab("Selected 3'UTR") + ylab("Relative contribution to gene TPM") + 
    theme(legend.position = "bottom") + 
    stat_summary(position = position_dodge(0.75),
                 fun.data = fun_length, geom = "text",
                 vjust = 0, size = 3) + 
    expand_limits(y = 1.25)
})
print(cowplot::plot_grid(
  cowplot::plot_grid(plotlist = lapply(plots, function(p) p + theme(legend.position = "none")), nrow = 2),
  cowplot::get_legend(plots[[1]]), ncol = 1, rel_heights = c(1, 0.05))
)
dev.off()

## Plot scores for modified and unmodified genes
genes <- scores$genes %>% 
  dplyr::mutate(geneclass = ifelse(gene %in% modgenes, "modified", "unmodified")) %>%
  dplyr::filter(uniqjuncreads > uniqjuncreadsthreshold)
png(gsub("\\.rds$", "_scores.png", outrds), width = 10, height = 7, unit = "in", res = 300)
print(ggplot(genes, aes(x = geneclass, y = score, color = geneclass)) + 
        geom_boxplot(outlier.size = 0.3) + 
        facet_wrap(~ method, nrow = 2) + theme_bw() + 
        scale_color_manual(values = c(modified = "#9900cc", unmodified = "#009933"),
                           name = "Gene class") + 
        xlab("") + theme(legend.position = "bottom"))
dev.off()

png(gsub("\\.rds$", "_scores2.png", outrds), width = 10, height = 7, unit = "in", res = 300)
print(ggplot(genes %>% dplyr::left_join(gene_summary %>% 
                                          dplyr::select(gene, method, utr_fraction_of_length) %>% 
                                          dplyr::distinct()),
             aes(x = utr_fraction_of_length, y = score)) + 
        geom_point(alpha = 0.3, size = 0.3) + geom_smooth() +  
        facet_wrap(~ method, nrow = 2) + theme_bw() + 
        xlab("Fraction of artificial transcript consisting of 3'UTR"))
dev.off()


saveRDS(list(genes = genes, gene_summary = gene_summary), file = outrds)
date()
sessionInfo()
