################################################################################
##                                                                            ##
## Plot correlation between scores from different methods                     ##
##                                                                            ##
## Inputs:                                                                    ##
## * scorerds: list containing abundance estimates and characteristics for    ##
##             junctions, transcripts and genes, as well as gene scores.      ##
##             Generated by calculate_gene_scores.R                           ##
## * quantmethods: string containing the quantification methods to consider,  ##
##                 separated by commas (no spaces)                            ##
## * uniqjuncreadsthreshold: the total number of uniquely mapping junction    ##
##                           reads (in a gene), only genes with more than     ##
##                           this number will be used for the comparison      ##
## * outrds: output rds file. The name will be used to determine the name of  ##
##           the output figures.                                              ##
##                                                                            ##
## Outputs:                                                                   ##
## * A png figure with a pairs plot showing the association between the       ##
##   from each pair of methods.                                               ##
## * A png figure with a dendrogram illustrating the similarities between the ##
##   methods                                                                  ##
##                                                                            ##
################################################################################

args <- (commandArgs(trailingOnly = TRUE))
for (i in 1:length(args)) {
  eval(parse(text = args[[i]]))
}

quantmethods <- strsplit(quantmethods, ",")[[1]]

print(scorerds)
print(quantmethods)
print(uniqjuncreadsthreshold)
print(outrds)

suppressPackageStartupMessages({
  library(dplyr)
  library(ggplot2)
  library(cowplot)
  library(GGally)
  library(pheatmap)
})

source("Rscripts/helper_plot_functions.R")

## Read scores
scores <- readRDS(scorerds)$genes

## Filter
highexpression <- scores %>% dplyr::filter(uniqjuncreads >= uniqjuncreadsthreshold) %>% 
  dplyr::pull(gene)
scores_wide <- scores %>% dplyr::filter(method %in% quantmethods & 
                                          gene %in% highexpression) %>%
  dplyr::select(gene, method, score) %>% 
  tidyr::spread(method, score) %>% as.data.frame() %>% tibble::column_to_rownames("gene")

## Pairs plot
points_identityline <- function(data, mapping, ...){
  ggplot(data = data, mapping = mapping) +
    geom_abline(intercept = 0, slope = 1) + 
    geom_point(alpha = 0.3, size = 0.25)
}
png(gsub("\\.rds$", "_pairs.png", outrds), width = 10, height = 10, 
    unit = "in", res = 300)
print(ggpairs(scores_wide,
              lower = list(continuous = points_identityline),
              upper = list(continuous = combinecor)) + 
        theme_bw() + xlab("JCC score") + ylab("JCC score"))
dev.off()

## Heatmap of mean of differences between each pair of methods
avediff <- matrix(0, ncol(scores_wide), ncol(scores_wide))
rownames(avediff) <- colnames(avediff) <- colnames(scores_wide)
for (rn in rownames(avediff)) {
  for (cn in rownames(avediff)) {
    avediff[rn, cn] <- mean(scores_wide[[rn]] - scores_wide[[cn]], na.rm = TRUE)
  }
}
png(gsub("\\.rds$", "_diffheatmap.png", outrds), width = 6, height = 6, 
    unit = "in", res = 300)
pheatmap::pheatmap(avediff, cluster_cols = FALSE, cluster_rows = FALSE, 
                   display_numbers = TRUE, number_format = "%.3f", 
                   color = colorRampPalette(colors = c("blue", "white", "red"))(30))
dev.off()

## Dendrogram
png(gsub("\\.rds$", "_clustering.png", outrds), width = 5, height = 5, 
    unit = "in", res = 300)
cormat <- cor(scores_wide, use = "pairwise.complete.obs")
distmat <- as.dist(sqrt(2*(1 - cormat)))
hcl <- hclust(distmat)
nodePar <- list(pch = c(NA, 19), col = "black")
par(mar = c(7.1, 4.1, 4.1, 2.1))
plot(as.dendrogram(hcl), type = "rectangle", ylab = "Dissimilarity", nodePar = nodePar)
dev.off()

saveRDS(NULL, file = outrds)
date()
sessionInfo()
